train_decoder.py

```diff
@@ -0,0 +1,462 @@
+    num_val = len(available_shards) - num_train - num_test
+    assert num_train + num_test + num_val == len(available_shards), f"{num_train} + {num_test} + {num_val} = {num_train + num_test + num_val} != {len(available_shards)}"
+    train_split, test_split, val_split = torch.utils.data.random_split(available_shards, [num_train, num_test, num_val], generator=torch.Generator().manual_seed(0))
+
+    # The shard number in the webdataset file names has a fixed width. We zero pad the shard numbers so they correspond to a filename.
+    train_urls = [webdataset_base_url.format(str(shard).zfill(shard_width)) for shard in train_split]
+    test_urls = [webdataset_base_url.format(str(shard).zfill(shard_width)) for shard in test_split]
+    val_urls = [webdataset_base_url.format(str(shard).zfill(shard_width)) for shard in val_split]
+    
+    create_dataloader = lambda tar_urls, shuffle, with_text=False, is_low=False: create_image_embedding_dataloader(
+        tar_url=tar_urls,
+        num_workers=num_workers,
+        batch_size=batch_size if not is_low else 32,
+        embeddings_url=embeddings_url,
+        index_width=index_width,
+        shuffle_num = None,
+        extra_keys= ["txt"] if with_text else [],
+        shuffle_shards = shuffle,
+        resample_shards = False, 
+        img_preproc=img_preproc,
+        handler=wds.handlers.warn_and_continue
+    )
+
+    train_dataloader = create_dataloader(train_urls, shuffle_train)
+    val_dataloader = create_dataloader(val_urls, shuffle_val_test, with_text=True)
+    test_dataloader = create_dataloader(test_urls, shuffle_val_test, with_text=True, is_low=True)
+    return {
+        "train": train_dataloader,
+        "val": val_dataloader,
+        "test": test_dataloader
+    }
+
+
+def create_decoder(device, decoder_config, unets_config):
+    """Creates a sample decoder"""
+    unets = []
+    for i in range(0, len(unets_config)):
+        unets.append(Unet(
+            **unets_config[i]
+        ))
+        unets[i].to(device=device)
+    
+    decoder = Decoder(
+        unet=tuple(unets),  # Must be tuple because of cast_tuple
+        **decoder_config
+    )
+    decoder.to(device=device)
+
+    return decoder
+
+def generate_samples(decoder, dataloader, epoch, device, step, n=5, text_prepend=""):
+    """
+    Generates n samples from the decoder and uploads them to wandb
+    Consistently uses the first n image embeddings from the dataloader
+    If the dataloader returns an extra text key, that is used as the caption
+    """
+    test_iter = iter(dataloader)
+    images = []
+    with torch.no_grad():
+        decoder.eval()
+        for data in test_iter:
+            if len(data) == 3:
+                img, emb, txt = data
+            else:
+                img, emb = data
+                txt = [""] * emb.shape[0]
+            # If the number of images is larger than n, we can remove all embeddings after the first n to speed up inference
+            if img.shape[0] > n:
```

Comment:
why are we doing this instead of just having the iterator return it in batches of `n`

Reply:
This is so that the existing training set can also be sampled from without always using the full batch size as that is very slow. A separate loader could be created for the same purpose I suppose.

Reply:
@Veldrovive ohh yup, let's use a separate dataloader with the proper `sampling_batch_size`

Reply:
I updated this so that the sampling code uses separate data loaders with a config that ensures consistency.